### PR TITLE
OpenZFS zdb changes

### PR DESCRIPTION
**Pull request by datacore-skumar**

### CODE REVIEW

File: include/sys/zfs_context.h [hunk 1]
```diff
@@ -419,6 +419,7 @@ void procfs_list_add(procfs_list_t *procfs_list, void *p);
 #define	kmem_cache_destroy(_c)	umem_cache_destroy(_c)
 #define	kmem_cache_alloc(_c, _f) umem_cache_alloc(_c, _f)
 #define	kmem_cache_free(_c, _b)	umem_cache_free(_c, _b)
+#define kmem_cache_free_aligned(_c, _b)	umem_cache_free_aligned(_c, _b)
 #define	kmem_debugging()	0
```
Comment (lundman) [inline include/sys/zfs_context.h:423]: 👍

File: include/sys/zfs_file.h [hunk 2]
```diff
@@ -22,7 +22,9 @@
 #ifndef	_SYS_ZFS_FILE_H
 #define	_SYS_ZFS_FILE_H
 
-#ifndef _KERNEL
+#ifndef _KERNEL && defined(_WIN32)
+typedef HANDLE zfs_file_t;
```
Comment (lundman) [inline include/sys/zfs_file.h:26]: ah interesting .. but hopefully wont be needed.

File: lib/libspl/include/umem.h [hunk 3]
```diff
@@ -134,10 +134,11 @@ umem_free(void* ptr, size_t size __maybe_unused)
     free(ptr);
 }
 
+
 static inline void
-umem_free_aligned(void* ptr, size_t size __maybe_unused)
+umem_free_aligned(void* ptr, size_t size)
 {
-    posix_memalign_free(ptr);
+    _aligned_free(ptr);
```
Comment (lundman) [inline lib/libspl/include/umem.h:141]: I think I invented `posix_memaligned_free()` for Windows, which just calls _aligned_free().  and it should be just "free" for the other platforms. That is to match the `posix_memalign()` call.

File: lib/libspl/include/umem.h [hunk 4]
```diff
@@ -205,6 +206,15 @@ umem_cache_free(umem_cache_t *cp, void *ptr)
...
 
+static inline void
+umem_cache_free_aligned(umem_cache_t* cp, void* ptr)
+{
+    if (cp->cache_destructor)
+	cp->cache_destructor(ptr, cp->cache_private);
+
+    umem_free_aligned(ptr, cp->cache_bufsize);
+}
+
```
Comment (lundman) [inline lib/libspl/include/umem.h:217]: 👍

File: lib/libzpool/kernel.c [hunk 5]
```diff
@@ -46,6 +46,7 @@
 #include <sys/zvol.h>
 #include <zfs_fletcher.h>
 #include <zlib.h>
+#include <winternl.h>
```
Comment (lundman) [inline lib/libzpool/kernel.c:49]: and wosix.h, and wrap in #ifdef _WIN32

File: lib/libzpool/kernel.c [hunk 6]
```diff
@@ -818,9 +818,8 @@ random_get_pseudo_bytes(uint8_t *ptr, size_t len)
 int
 ddi_strtoul(const char *hw_serial, char **nptr, int base, unsigned long *result)
 {
-	char *end;
-
-	*result = strtoul(hw_serial, &end, base);
+	char* end;
+	*result = strtoul(hw_serial, nptr, base);
```
Comment (lundman) [inline lib/libzpool/kernel.c:822]: Pleasing you found this as well, sorry I didn't see it before fixing it. Other ZFS code can pass NULL as result, and Win32 will abort on it, so it needed slightly different fix,

File: lib/libzpool/kernel.c [hunk 7]
```diff
@@ -1090,64 +1089,47 @@ zvol_rename_minors(spa_t *spa, const char *oldname, const char *newname,
...
-	}
-
-	(void) fcntl(fd, F_SETFD, FD_CLOEXEC);
-
-	fp = umem_zalloc(sizeof (zfs_file_t), UMEM_NOFAIL);
-	fp->f_fd = fd;
-	fp->f_dump_fd = dump_fd;
-	*fpp = fp;
-
-	return (0);
```
Comment (lundman) [inline lib/libzpool/kernel.c:1140]: Hoping wosix.h will make all this working without much change, and any additional fixes to go into our posix.c

File: module/os/linux/zfs/abd_os.c [hunk 8]
```diff
@@ -583,7 +583,7 @@ abd_free_chunks(abd_t *abd)
 	abd_for_each_sg(abd, sg, n, i) {
 		for (int j = 0; j < sg->length; j += PAGESIZE) {
 			struct page *p = nth_page(sg_page(sg), j >> PAGE_SHIFT);
-			umem_free(p, PAGESIZE);
+			umem_free_aligned(p, PAGESIZE);
```
Comment (lundman) [inline module/os/linux/zfs/abd_os.c:586]: we don't have to change other platform files, but it does make it cleaner....

File: module/zfs/spa_config.c [hunk 9]
```diff
@@ -96,14 +97,24 @@ spa_config_load(void)
...
+	if (!strncmp(spa_config_path, "\\SystemRoot\\", 12)) {
+	    (void)snprintf(pathname, MAXPATHLEN, "%s\\%s",
+		getenv("SystemRoot"), spa_config_path + 12);
+	}
+	else {
+	    (void)snprintf(pathname, MAXPATHLEN, "%s%s",
+		"", spa_config_path);
+	}
+#endif
+	err = zfs_file_open(pathname, O_RDONLY, 0, &fp);
```
Comment (lundman) [inline module/zfs/spa_config.c:117]: lets move this check into our `zfs_file_open` if we can

**Comment by lundman** [comment]
I took the remaining bits out of this PR and put it into zdb. Basic usage works, and we need to go through all the options and find other cases that fails. Thank you for your work, even if it didn't go in the usual way.  We still need to work on fixing import, but zdb -e can perhaps help with that, since you can easily see the import logic prints

**Comment by datacore-skumar** [comment]
```
Command input : zdb -h tank12345
Output : wosix_open(\\?\PHYSICALDRIVE1): error 32 / 0x20
         zdb: can't open 'tank12345': Resource device
```
```
0:413> k 
# Child-SP          RetAddr           Call Site
00 0000004a`11aff520 00007ff7`5d87e5aa zdb!wosix_open+0x5cd [C:\Users\skumar\Desktop\OpenZFS_Jorgen\openzfs\lib\libspl\os\windows\posix.c @ 964]
01 0000004a`11affb00 00007ff7`5d3880fc zdb!wosix_stat+0x1a [C:\Users\skumar\Desktop\OpenZFS_Jorgen\openzfs\lib\libspl\os\windows\posix.c @ 1186]
02 0000004a`11affb50 00007ff7`5d58d9b6 zdb!zfs_file_open+0x6c [C:\Users\skumar\Desktop\OpenZFS_Jorgen\openzfs\lib\libzpool\kernel.c @ 1103]
03 0000004a`11affc20 00007ff7`5d493683 zdb!vdev_file_open+0x3c6 [C:\Users\skumar\Desktop\OpenZFS_Jorgen\openzfs\module\os\windows\zfs\vdev_file.c @ 172]
04 0000004a`11affcd0 00007ff7`5d49425e zdb!vdev_open+0x3e3 [C:\Users\skumar\Desktop\OpenZFS_Jorgen\openzfs\module\zfs\vdev.c @ 1749]
05 0000004a`11affda0 00007ff7`5d55e17f zdb!vdev_open_child+0x2e [C:\Users\skumar\Desktop\OpenZFS_Jorgen\openzfs\module\zfs\vdev.c @ 1606]
06 0000004a`11affde0 00007ff7`5d388a4c zdb!taskq_thread+0x1cf [C:\Users\skumar\Desktop\OpenZFS_Jorgen\openzfs\lib\libzpool\taskq.c @ 238]
07 0000004a`11affe50 00007ff7`5d9febcc zdb!pthread_create_wrapper+0x7c [C:\Users\skumar\Desktop\OpenZFS_Jorgen\openzfs\lib\os\windows\libpthread\pthread.h @ 787]
08 0000004a`11affea0 00007ffd`39c57974 zdb!thread_start<unsigned int (__cdecl*)(void *),1>+0x9c [minkernel\crts\ucrt\src\appcrt\startup\thread.cpp @ 97]
09 0000004a`11afff00 00007ffd`3bf5a271 KERNEL32!BaseThreadInitThunk+0x14
0a 0000004a`11afff30 00000000`00000000 ntdll!RtlUserThreadStart+0x21
```

**Comment by lundman** [comment]
Ah yes, me too - any idea why it complains about it? As far as I can see, we close all handles, and everything is opened share_read.   I can reboot, and it works again, for a little while...

**Comment by datacore-skumar** [comment]
Closed this PR because some of the changes are already merged by Jorgen.

**Pull request closed without merging**